server/filestore.go

```diff
@@ -933,52 +935,67 @@ func (mb *msgBlock) convertCipher() error {
 	if len(ekey) < minBlkKeySize {
 		return errBadKeySize
 	}
-	// Recover key encryption key.
-	rb, err := fs.prf([]byte(fmt.Sprintf("%s:%d", fs.cfg.Name, mb.index)))
-	if err != nil {
-		return err
+	type prfWithCipher struct {
+		keyGen
+		StoreCipher
 	}
-	kek, err := genEncryptionKey(osc, rb)
-	if err != nil {
-		return err
+	var prfs []prfWithCipher
+	if fs.prf != nil {
+		prfs = append(prfs, prfWithCipher{fs.prf, sc})
+		prfs = append(prfs, prfWithCipher{fs.prf, osc})
 	}
-	ns := kek.NonceSize()
-	seed, err := kek.Open(nil, ekey[:ns], ekey[ns:], nil)
-	if err != nil {
-		return err
+	if fs.oldprf != nil {
+		prfs = append(prfs, prfWithCipher{fs.oldprf, sc})
+		prfs = append(prfs, prfWithCipher{fs.oldprf, osc})
 	}
-	nonce := ekey[:ns]
 
-	bek, err := genBlockEncryptionKey(osc, seed, nonce)
-	if err != nil {
-		return err
-	}
+	for _, prf := range prfs {
+		// Recover key encryption key.
+		rb, err := prf.keyGen([]byte(fmt.Sprintf("%s:%d", fs.cfg.Name, mb.index)))
+		if err != nil {
+			continue
+		}
+		kek, err := genEncryptionKey(prf.StoreCipher, rb)
+		if err != nil {
+			continue
+		}
+		ns := kek.NonceSize()
+		seed, err := kek.Open(nil, ekey[:ns], ekey[ns:], nil)
+		if err != nil {
+			continue
+		}
+		nonce := ekey[:ns]
+		bek, err := genBlockEncryptionKey(prf.StoreCipher, seed, nonce)
+		if err != nil {
+			return err
+		}
 
-	buf, _ := mb.loadBlock(nil)
-	bek.XORKeyStream(buf, buf)
-	// Make sure we can parse with old cipher and key file.
-	if err = mb.indexCacheBuf(buf); err != nil {
-		return err
-	}
-	// Reset the cache since we just read everything in.
-	mb.cache = nil
+		buf, _ := mb.loadBlock(nil)
+		bek.XORKeyStream(buf, buf)
+		// Make sure we can parse with old cipher and key file.
+		if err = mb.indexCacheBuf(buf); err != nil {
+			return err
+		}
+		// Reset the cache since we just read everything in.
+		mb.cache = nil
 
-	// Generate new keys based on our
-	if err := fs.genEncryptionKeysForBlock(mb); err != nil {
-		// Put the old keyfile back.
-		keyFile := filepath.Join(mdir, fmt.Sprintf(keyScan, mb.index))
-		os.WriteFile(keyFile, ekey, defaultFilePerms)
-		return err
-	}
-	mb.bek.XORKeyStream(buf, buf)
-	if err := os.WriteFile(mb.mfn, buf, defaultFilePerms); err != nil {
-		return err
+		// Generate new keys based on our
```

Comment:
Finish comment?

Reply:
Funnily enough this unfinished comment was there before, I'll update it.